### PR TITLE
Add accounts search endpoint

### PR DESCRIPTION
**Pull request by samuelhwilliams**
### Change description
On post-award we need a way to list all accounts with the (new) SECTION_151 role that are under a specific email domain.

This new endpoint provides that core functionality.

https://dluhcdigital.atlassian.net/browse/FPASF-477

- [x] Unit tests and other appropriate tests added or updated
- [x] README and other documentation has been updated / added (if needed)
- [x] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
* Enable the swagger UI by updating `app.py` to:
    ```
    connexion_app = connexion.FlaskApp(
        __name__,
        specification_dir=Config.FLASK_ROOT + "/openapi/",
        swagger_ui_options=SwaggerUIOptions(swagger_ui_path="/docs"),
    )
    ```
* Run the app and go to http://localhost:3003/docs/#/accounts/core.account.search_accounts
* Use the swagger UI to test the new endpoint.

### CODE REVIEW

File: openapi/api.yml [hunk 1]
```diff
@@ -210,6 +230,29 @@ components:
...
+          type: array
+          items:
+            type: string
+        partial_roles:
+          description: "Filter results to accounts that have roles partially matching ANY of these values."
+          example: ['ASSESSOR']
+          type: array
+          items:
+            type: string
+      not:
```
Comment (MarcUsher) [inline openapi/api.yml:254]: Again, asking questions here for my own education - here we're saying `roles` and `partial_roles` aren't required, but we haven't said that `email_domain` is required so technically they're all optional, is that right? Is there a reason we're doing it this way rather than saying eg. `email_domain` is required?

File: openapi/api.yml [hunk 2]
```diff
@@ -149,6 +149,26 @@ paths:
...
+      tags:
+        - accounts
+      summary: Search for accounts based on certain criteria
+      operationId: core.account.search_accounts
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/accountSearch'
+      responses:
```
Comment (MarcUsher) [inline openapi/api.yml:163]: Related to the other comment about required fields, do we ever need to account for the endpoint returning no results? If no args are passed and no filters applied we'd get all the accounts back, but there's potentially a combo of filters (an email domain and a role) that could return nothing if there's no one with that role and email domain? And also some of the tests are checking for things that return 0 results, just wondering if that should still be a 200 or handled differently.

**Review by MarcUsher — APPROVED** [review]
Approving, again with 2 small comments that's more for me to understand properly! Thanks!

File: openapi/api.yml (hunk 1 above)
Comment (samuelhwilliams) [reply openapi/api.yml:254]: All three fields are optional.

The funkiness here is saying that we don't accept any requests that provide _both_ `roles` and `partial_roles`. It doesn't read super clearly using the wording here. `not` followed by `required` says "these things cannot all be present", rather than "these two things are not required" 🙃

Comment (MarcUsher) [reply openapi/api.yml:254]: Gotcha! Thanks!

File: openapi/api.yml (hunk 2 above)
Comment (samuelhwilliams) [reply openapi/api.yml:163]: It is valid for a search to return no results - the answer to the question is just an empty list. It doesn't make sense to return a 404 in that case - that's more about saying the resource doesn't exist. In this case the resource is a search (result) - which you will always get.

In my opinion, `get_accounts_for_fund` is doing the wrong thing by returning a 404 for their search 🙃 🤐

Comment (MarcUsher) [reply openapi/api.yml:163]: Fair point, thanks for clarifying!

File: openapi/api.yml [hunk 3]
```diff
@@ -150,6 +150,26 @@ paths:
                   $ref: '#/components/schemas/account'
         404:
           description: "No associated accounts found."
+  /accounts/search:
+    post:
```
Comment (albertkol) [inline openapi/api.yml:154]: Any particular reason why you opted to make this a POST request? Thought reading data fits better a GET request. In hindsight, it could have potentially be the `get_accounts` with some extra parameters.

Comment (samuelhwilliams) [reply openapi/api.yml:154]: Yeah it's a good question - normally idempotent requests would be a GET, you're right.

I prefer POST for 'search'-style requests because of a few generic reasons:

* they allow transmission of more data (before you hit weird things around how long URL query parameters can be)
* they prevent search terms being logged, especially if you're searching for specific email addresses (here we aren't - but maybe in the future this endpoint could)
* HTTP spec treats this as a single 'search' endpoint/resource - not a load of individual resources depending on what params are passed.

Basically - it's my preferred option over GET until we get something like [HTTP QUERY](https://httpwg.org/http-extensions/draft-ietf-httpbis-safe-method-w-body.html)

**Review by albertkol — APPROVED** [review]
:+1:
